src/rust/inetstack/protocols/tcp/peer.rs

```diff
@@ -270,25 +262,25 @@ impl SharedTcpPeer {
     }
 
     /// Processes an incoming TCP segment.
-    pub fn receive(&mut self, ip_hdr: Ipv4Header, buf: DemiBuffer) {
-        let (tcp_hdr, data): (TcpHeader, DemiBuffer) =
-            match TcpHeader::parse(&ip_hdr, buf, self.tcp_config.get_rx_checksum_offload()) {
-                Ok(result) => result,
-                Err(e) => {
-                    let cause: String = format!("invalid tcp header: {:?}", e);
-                    error!("receive(): {}", &cause);
-                    return;
-                },
-            };
+    pub fn receive(&mut self, src_ipv4_addr: Ipv4Addr, buf: DemiBuffer) {
+        // We can assume that the destination is our local IPv4 address; otherwise, the IP layer would have discarded
+        // the packet already.
+        let (tcp_hdr, data): (TcpHeader, DemiBuffer) = match TcpHeader::parse(
+            &src_ipv4_addr,
+            &self.local_ipv4_addr,
+            buf,
+            self.tcp_config.get_rx_checksum_offload(),
+        ) {
+            Ok(result) => result,
+            Err(e) => {
+                let cause: String = format!("invalid tcp header: {:?}", e);
+                error!("receive(): {}", &cause);
+                return;
+            },
+        };
         debug!("TCP received {:?}", tcp_hdr);
-        let local: SocketAddrV4 = SocketAddrV4::new(ip_hdr.get_dest_addr(), tcp_hdr.dst_port);
-        let remote: SocketAddrV4 = SocketAddrV4::new(ip_hdr.get_src_addr(), tcp_hdr.src_port);
-
-        if remote.ip().is_broadcast() || remote.ip().is_multicast() || remote.ip().is_unspecified() {
-            let cause: String = format!("invalid remote address (remote={})", remote.ip());
-            error!("receive(): {}", &cause);
-            return;
-        }
+        let local: SocketAddrV4 = SocketAddrV4::new(self.local_ipv4_addr, tcp_hdr.dst_port);
```

Comment:
Could rename these to `local_socket_addrv4` and `remote_socket_addrv4` becuase there are several addrs in this area of the codebase.

Reply:
I'm going to skip doing this because this code will be moved soon but will note it to fix later